http/org.wso2.carbon.transport.http.netty/src/test/java/org/wso2/carbon/transport/http/netty/websocket/HttpToWebSocketProtocolSwitchHttpListener.java

```diff
@@ -0,0 +1,73 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.transport.http.netty.websocket;
+
+import io.netty.handler.codec.http.HttpHeaders;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.transport.http.netty.common.Constants;
+import org.wso2.carbon.transport.http.netty.contract.HttpConnectorListener;
+import org.wso2.carbon.transport.http.netty.contract.ServerConnectorException;
+import org.wso2.carbon.transport.http.netty.message.HTTPCarbonMessage;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * WebSocket test class the Protocol switch from HTTP to WebSocket, WebSocket Listener.
+ */
+public class HttpToWebSocketProtocolSwitchHttpListener implements HttpConnectorListener {
+    private static final Logger logger = LoggerFactory.getLogger(HttpToWebSocketProtocolSwitchHttpListener.class);
+
+    private ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    public void onMessage(HTTPCarbonMessage httpRequest) {
+        executor.execute(() -> {
+            try {
+                int length = httpRequest.getFullMessageLength();
```

Comment:
Here we are doing something similar to cloning the HTTPCarbonMessage. Isn't it better to move this logic to a clone method within the transport implementation rather than writing this logic here at the test class? This functionality can be used even when ballerina user needs to convert HTTP to WebSocket within ballerina (not in test cases). WDYT?

Reply:
Added cloning methods to HTTPCarbonMessage and made necessary changes